crates/kubelet/src/handle.rs

```diff
@@ -33,27 +34,123 @@ pub trait Stop {
+/// dependency of `kubelet`.
+pub trait LogHandle<R>: Sync + Send {
+    /// Create new log reader.
+    fn output(&self) -> R;
+}
+
+/// Future that streams logs from provided `AsyncRead` to provided `hyper::body::Sender`.
+async fn stream_logs<R: AsyncRead + std::marker::Unpin>(
+    output: R,
+    mut sender: hyper::body::Sender,
+    tail: Option<usize>,
+    follow: bool,
+) -> anyhow::Result<()> {
+    let buf = tokio::io::BufReader::new(output);
+    let mut lines = buf.lines();
+
+    if let Some(n) = tail {
+        // Stream last n lines.
+        // TODO: this uses a lot of memory for large n and scans the entire file.
+        let mut line_buf = std::collections::VecDeque::with_capacity(n);
+
+        while let Some(line) = lines.next_line().await? {
+            if line_buf.len() == n {
+                line_buf.pop_front();
+            }
+            line_buf.push_back(line);
+        }
+
+        for mut line in line_buf {
+            line.push('\n');
+            let b = hyper::body::Bytes::copy_from_slice(&line.as_bytes());
+            match sender.send_data(b).await {
+                Ok(_) => (),
+                Err(e) => {
+                    if e.is_closed() {
+                        debug!("channel closed.");
+                        return Ok(());
+                    } else {
+                        error!("channel error: {}", e);
+                        bail!(e);
+                    }
+                }
+            }
+        }
+    } else {
+        // Stream entire file.
+        while let Some(mut line) = lines.next_line().await? {
+            line.push('\n');
+            let b = hyper::body::Bytes::copy_from_slice(&line.as_bytes());
+            match sender.send_data(b).await {
+                Ok(_) => (),
+                Err(e) => {
+                    if e.is_closed() {
+                        debug!("channel closed.");
+                        return Ok(());
+                    } else {
+                        error!("channel error: {}", e);
+                        bail!(e);
+                    }
+                }
+            }
+        }
+    }
+
+    if follow {
+        // Optionally watch file for changes.
+        loop {
+            while let Some(mut line) = lines.next_line().await? {
+                line.push('\n');
+                let b = hyper::body::Bytes::copy_from_slice(&line.as_bytes());
+                match sender.send_data(b).await {
+                    Ok(_) => (),
+                    Err(e) => {
+                        if e.is_closed() {
+                            debug!("channel closed.");
+                            return Ok(());
+                        } else {
+                            error!("channel error: {}", e);
+                            bail!(e);
+                        }
+                    }
+                }
+            }
+            tokio::time::delay_for(std::time::Duration::from_millis(500)).await;
```

Comment:
It doesn't block at EOF I believe (notice that I don't have to detect EOF when streaming the file at the start of the function. So I think this will busy wait on the outside loop without this delay.